### PR TITLE
more correct implementation of implicit regressor 

### LINKED ISSUE #672: experiment with InfoNCE loss instead of BCE loss for implicit regression
as mentioned here https://arxiv.org/pdf/2109.00137.pdf

### PR DESCRIPTION
**Pull request by tomsilver**
closes #672 

on the first 5 seeds, `python src/main.py --approach nsrt_learning --env touch_point --option_learner implicit_bc` now gets 50/50 in all cases (previously, one seed would get 49).

visualization with 50 demos (note: unlike before, the colors are now energies, not probabilities):
touch_point_learned_energy

visualization with 500 demos:
touch_point_learned_energy

a few observations:
* the difference between 50 and 500 is no longer as obvious. but the results with 50 look good.
* the NN apparently hasn't figured out that -np.pi == np.pi in this action space, as evidenced by the discontinuity in the color map on the left. I'm not sure what to think about that... maybe it's related to the fact that we're only training on expert data with no noise in it.

### CODE REVIEW

File: src/torch_models.py [hunk 1]
```diff
@@ -364,26 +363,43 @@ def _initialize_net(self) -> None:
...
-            # Convert data to tensors.
-            tensor_X = torch.from_numpy(
-                np.array(concat_inputs, dtype=np.float32))
-            yield (tensor_X, tensor_Y)
+            # Resample negative examples on each iteration.
+            neg_Y = torch.rand(size=(num_samples, num_negatives, self._y_dim),
+                               dtype=tensor_Y.dtype)
+            # Create a multiclass classification-style target vector.
+            combined_Y = torch.cat([tensor_Y, neg_Y], axis=1)  # type: ignore
+            combined_Y = combined_Y.reshape([-1, tensor_Y.shape[-1]])
```
Comment (ronuchit) [inline src/torch_models.py:393]: i don't understand what this line is doing. can you go through a concrete example of the shapes?
so tensor_X is (num_samples, state_dim)
and tensor_Y is (num_samples, action_dim)
we are creating num_neg_samples negative examples of actions, so combined_Y should be (num_samples + num_neg_samples, action_dim)
then what's the reshape doing? and why is the `neg_Y = ...` 3-dimensional? i'm clearly misunderstanding something...

File: src/torch_models.py [hunk 2]
```diff
@@ -364,26 +363,43 @@ def _initialize_net(self) -> None:
         self._linears.append(nn.Linear(self._hid_sizes[-1], 1))
 
     def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
-        return nn.BCEWithLogitsLoss()
+
+        def _loss_fn(Y_hat: Tensor, Y: Tensor) -> Tensor:
+            pred = Y_hat.reshape(Y.shape)  # (num data, num classes)
```
Comment (ronuchit) [inline src/torch_models.py:368]: why is Y_hat not already the shape of Y?

**Review by ronuchit — APPROVED** [review]
just generally very confused on the shapes of things but i trust that it's correct based on the results

File: src/torch_models.py [hunk 3]
```diff
@@ -364,26 +363,43 @@ def _initialize_net(self) -> None:
...
+            log_probs = F.log_softmax(pred / self._temperature, dim=-1)
+            loss = F.kl_div(log_probs, Y, reduction='batchmean')
```
Comment (ronuchit) [inline src/torch_models.py:370]: why this setting of reduction?

File: src/torch_models.py [hunk 4]
```diff
@@ -364,26 +363,43 @@ def _initialize_net(self) -> None:
...
+            combined_Y = torch.cat([tensor_Y, neg_Y], axis=1)  # type: ignore
+            combined_Y = combined_Y.reshape([-1, tensor_Y.shape[-1]])
+            # Concatenate to create the final input to the network.
+            XY = torch.cat([extended_X, combined_Y], axis=1)  # type: ignore
+            assert XY.shape == ((num_negatives + 1) * num_samples,
+                                self._x_dim + self._y_dim)
+            # Create labels for multiclass loss. Note that the true inputs
+            # are first, so the target labels are all zeros.
+            indices = torch.zeros([num_samples], dtype=torch.int64)
+            labels = F.one_hot(indices, num_classes=num_negatives + 1).float()
```
Comment (ronuchit) [inline src/torch_models.py:401]: waitttt, why aren't there just 2 classes? positive and negative?

File: src/torch_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomsilver) [reply src/torch_models.py:368]: it's recommended here: https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html#torch.nn.KLDivLoss adding a comment

File: src/torch_models.py (hunk 3 above)
Comment (tomsilver) [reply src/torch_models.py:370]: it's recommended here: https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html#torch.nn.KLDivLoss adding a comment

File: src/torch_models.py (hunk 4 above)
Comment (tomsilver) [reply src/torch_models.py:401]: adding a docstring comment to hopefully clarify

File: src/torch_models.py [hunk 5]
```diff
@@ -364,26 +363,43 @@ def _initialize_net(self) -> None:
...
-        targets = np.array([1] * num_pos_inputs + [0] * num_neg_inputs,
-                           dtype=np.float32)
-        tensor_Y = torch.from_numpy(np.array(targets, dtype=np.float32))
+        num_samples = X.shape[0]
+        num_negatives = self._num_negatives_per_input
+        # Cast to torch first.
+        tensor_X = torch.from_numpy(np.array(X, dtype=np.float32))
+        tensor_Y = torch.from_numpy(np.array(Y, dtype=np.float32))
+        # Expand tensor_Y in preparation for concat in the loop below.
+        tensor_Y = tensor_Y[:, None, :]
```
[the remark this answers is not in the document]
Comment (tomsilver) [reply src/torch_models.py:383]: this is pytorch!

**Comment by tomsilver** [comment]
@ronuchit I added some comments and asserts to try to address your questions. can you take another look at let me know which parts I should try to further clarify? (sorry, this is hard... I implemented it this way mainly by looking at references)

File: src/torch_models.py [hunk 6]
```diff
@@ -319,13 +319,32 @@ def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
...
+        L = E[-log(p(y | x, {y'}))]
+
+        p(y | x, {y'})) = exp(-f(x, y)) / [
+            (exp(-f(x, y)) + sum_{y'} exp(-f(x, y')))
+        ]
+
+    where (x, y) is an example "positive" input/output from (X, Y), f is
+    the model that we are learning in this class, and {y'} is a set of
+    "negative" output examples for input x. The size of that set is
+    self._num_negatives_per_input (num_negatives).
```
Comment (ronuchit) [inline src/torch_models.py:340]: what does the `(num_negatives)` mean here -- i thought this sentence is sufficiently clear without it

File: src/torch_models.py [hunk 7]
```diff
@@ -319,13 +319,32 @@ def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
...
+    loss that looks like this:
+
+        L = E[-log(p(y | x, {y'}))]
+
+        p(y | x, {y'})) = exp(-f(x, y)) / [
+            (exp(-f(x, y)) + sum_{y'} exp(-f(x, y')))
+        ]
+
+    where (x, y) is an example "positive" input/output from (X, Y), f is
+    the model that we are learning in this class, and {y'} is a set of
```
Comment (ronuchit) [inline src/torch_models.py:338]: model -> energy function

File: src/torch_models.py [hunk 8]
```diff
@@ -319,13 +319,32 @@ def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
...
+    "negative" output examples for input x. The size of that set is
+    self._num_negatives_per_input (num_negatives).
+
+    One way to interpret the expression is that the numerator exp(-f(x, y))
```
Comment (ronuchit) [inline src/torch_models.py:342]: smart! did you come up with this, or is this normally how people implement it?

File: src/torch_models.py [hunk 9]
```diff
@@ -319,13 +319,32 @@ def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
...
+    represents an unnormalized probability that the same (x, y) belongs to
+    a certain ground truth "class". Each of the exp(-f(x, y') in the
```
Comment (ronuchit) [inline src/torch_models.py:344]: missing a `)`

File: src/torch_models.py [hunk 10]
```diff
@@ -319,13 +319,32 @@ def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
...
+            (exp(-f(x, y)) + sum_{y'} exp(-f(x, y')))
+        ]
+
+    where (x, y) is an example "positive" input/output from (X, Y), f is
+    the model that we are learning in this class, and {y'} is a set of
+    "negative" output examples for input x. The size of that set is
+    self._num_negatives_per_input (num_negatives).
+
+    One way to interpret the expression is that the numerator exp(-f(x, y))
+    represents an unnormalized probability that the same (x, y) belongs to
```
Comment (ronuchit) [inline src/torch_models.py:343]: what do you mean by `the same (x, y)`? maybe just `this (x, y)`?

File: src/torch_models.py [hunk 11]
```diff
@@ -364,26 +384,60 @@ def _initialize_net(self) -> None:
...
     def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
-        return nn.BCEWithLogitsLoss()
+
+        # See the class docstring for context.
+        def _loss_fn(Y_hat: Tensor, Y: Tensor) -> Tensor:
+            pred = Y_hat.reshape(Y.shape)  # (num data, num classes)
+            log_probs = F.log_softmax(pred / self._temperature, dim=-1)
+            # Note: batchmean is recommended in the PyTorch documentation
+            # and will become the default in a future version.
+            loss = F.kl_div(log_probs, Y, reduction='batchmean')
```
Comment (ronuchit) [inline src/torch_models.py:394]: i thought the loss would be a multiclass classification loss like you were saying above. is this equivalent?

File: src/torch_models.py [hunk 12]
```diff
@@ -364,26 +384,60 @@ def _initialize_net(self) -> None:
...
+                                self._x_dim + self._y_dim)
+            # Create labels for multiclass loss. Note that the true inputs
+            # are first, so the target labels are all zeros.
+            indices = torch.zeros([num_samples], dtype=torch.int64)
+            labels = F.one_hot(indices, num_classes=num_negatives + 1).float()
+            assert labels.shape == (num_samples, num_negatives + 1)
+            # Note that XY is flattened and labels is not. XY is flattened
+            # because we need to feed each entry through the network during
+            # training. Labels is unflattened because we will want to use
+            # KLDivLoss.
```
Comment (ronuchit) [inline src/torch_models.py:439]: i don't see KLDivLoss anywhere? do you mean F.kl_div?

**Review by ronuchit — APPROVED** [review]
lots of questions, but unblocking

File: src/torch_models.py (hunk 8 above)
Comment (tomsilver) [reply src/torch_models.py:342]: I did not come up with it! I'm heavily relying on a private reference implementation.

File: src/torch_models.py (hunk 11 above)
Comment (tomsilver) [reply src/torch_models.py:394]: yeah, I think they should be equivalent: https://stats.stackexchange.com/questions/357963/what-is-the-difference-cross-entropy-and-kl-divergence

(I'm opting for KL divergence just in case, because of the reference implementation that I'm following.)

to your second question, I'll add more comments in the code
